Documentation/design-docs/host-testing.md

```diff
@@ -0,0 +1,102 @@
+Ideally we would migrate these over time to the proposed Component tests infra.
+
+Pros:
+* Testing real product binaries
+* Tests are written in C#, but in a separate project - no direct test infra integration or VS UI support. This is worked around by effectively having "wrappers" for these tests.
+* Direct calls to exports in a controlled environment without running the whole product (mostly)
+
+Cons:
+* Still runs out-of-proc making debugging harder
+
+## Test proposal
+
+### Unit tests
+Add the ability to write true unit tests. So tests are written in C++ which can directly call methods/classes from the product. These tests would be compiled as native executables. For simplicity of the build system, they would live in the product source tree (`src` folder) and would be compiled during the product build. They would combine the test source files with product source files into one executable.
+For integration purposes there would be managed wrappers in the `HostActivation` test project which would just call the native executable and check the exit code. As this is the simplest way to integrate these tests into all our test infrastructure.
+
+Unit tests should be used to test self-contained bits of functionality. So for example helper classes and such.
+
+First unit test is being introduced in this PR: [dotnet/core-setup#4953](https://github.com/dotnet/core-setup/pull/4953).
+
+Pros:
+* Can test C++ code directly
+* Can create very isolated environment for the tested code
+
+Cons:
+* Test must be written in C++
+* Not testing shipping binaries
+* Debugging is trickier as it launches a separate process (this could be fixed)
+
+### Component tests
+Testing larger functionality blocks is probably easier by testing entire components.
+
+These tests would call entry points on the host components directly (so `hostfxr` and `hostpolicy` mostly) and observe the resulting behavior. The tests would **not** try to emulate real product installation. Instead they would rely on mocks.
+The core of this proposed test approach is
+* The tested component (`hostfxr` for example) is loaded directly into the test process - so it would be running on some toolset version of .NET Core and not the tested version of .NET Core - but that should not matter as all of that should be abstracted out.
+* Mock file system access and other OS services which would otherwise require complex setup
+* Mock dependent libraries (so for example `coreclr`) as necessary
+
+Product changes would be required to allow for mocking OS services. This would be achieved by having test-only exports/env variables which would turn on the mocking. For example file system mocking would be done by abstracting file system access through and interface and having an export which would let the test provide custom implementation of that interface.
+There are two way to do this:
```

Comment:
The choice between he two options depends on how much of code the runtime/os abstraction layer is expected to be. My preference would be to keep the test code separate from the final released product. Even though this means testing two builds, I think much of the difference can be hidden under a common interface. For example:
* Using a different coreclr.dll with the interface unchanged, so that host code remains unchanged
* Have a PAL like abstraction (but much simpler than the PAL) that provides the file-operation interface (but provides mock answers in the test build).

Reply:
Actually my thinking was more in line to introduce interfaces in the product to create separation. That would remain in the code for good (the perf impact is negligible and it introduces nice separation of concern on its own).
The test-only bits would be some ability to provide a different implementation of such interfaces. The implementation itself would not be part of the product dll - that would come from the outside.

I definitely agree that mocking external dependencies (coreclr, hospolicy for hostfxr tests and so on) is a no brainer should be done.

I personally don't want to have testing support only in debug builds. All build flavors should be testable. The open question is if we want the shipping binaries to be testable like this, or of it's OK to have a second build of them designed for testing.

Reply:
I personally would slightly prefer to have just one build - and accept the presence of a test hook in the shipping product.

Reply:
Forcing a well defined interface that enables proper testing forces us to design a solution. Having a special build is a way to avoiding spending time in test.